### DISCUSSION TITLE
What is the best practice of customizing the logo of the login page?

### DISCUSSION THREAD

**Discussion by itsdapi** [keycloakify/keycloakify#561] [Ideas]
I noticed that the keycloak logo was baked into the theme. And all what KC provide is html elements input, not able to upload logo. 

So besides using online resources in inline css url, what is the best practice to do that? I don't want to compile a new theme if new logo comes in.

**Comment by garronej** [comment]
There are multiple strategies to acheive this.  

You have the approach based on environnement variable as used in the french governement Keycloak theme:  
https://github.com/codegouvfr/keycloak-theme-dsfr?tab=readme-ov-file#environment-variables  

You have also the dynamic theme customization with values that can be passed by the client (be carfull to enforce origins!) It's the approach implemented for Onyxia:  
https://youtu.be/NrVuVXsbloA?si=1DQiUBSqae0FFtCc&t=343  

Ther's also the possibility of [importing assets from the public directory](https://docs.keycloakify.dev/importing-assets) in your code. This enables you to hotswap the assets in your keycloak instance `/opt/keycloak/themes/<your-theme>/login/resources/build/`

You might aslo consider theme variant although it doesn't seem to fit your usecase exactly.  

Let me know know more about your usecase and I'll tell you what's the best option for you.

**Reply by itsdapi** [reply]
sorry for the latey reply. I think this is the kinda feature that kc need to provide or at least a plugin, not the theme itself xD.

Thanks for advice though. I might consider just using text or using selfhosted img source
